### PR TITLE
Stop large class masking jobs from being cancelled before they report any progress

### PR DESCRIPTION
**Pull request by mihow**
## Summary

Large class masking runs were being cancelled roughly eleven minutes in, having reported no progress at all: the job page showed zero classifications checked. The cause was not the masking work itself but the query that decides which classifications are in scope, which asked the database to de-duplicate rows it could not have duplicated. On a scope of 55,530 classifications that took 208 seconds to count and just as long again before the first row could be read, so the run crossed the stalled-job cutoff before it ever committed a batch and reported progress.

Removing the de-duplication brings the same count back in half a second. A masking run now starts working almost immediately and reports the size of the job before it processes the first prediction, so an operator watching a long run can see how much was found rather than a stage sitting at zero.

Checking that claim in review turned up a smaller related gap: post-processing jobs never marked their stage started, so they read as "Waiting to start" until the first progress report. That is fixed here too, and it affects the small size filter as well.

The scope query returns exactly the same classifications, and the masking logic is untouched.

### List of Changes

| # | What changes for the operator | How |
|---|---|---|
| 1 | Large masking runs are no longer cancelled before they begin | `ClassMaskingTask._scoped_classifications` no longer calls `.distinct()` on either scope |
| 2 | The `run_class_masking` command's preflight count returns promptly instead of appearing to hang | Same de-duplication removed from the command's count |
| 3 | A running job shows how many classifications it found before it starts working through them | New `on_setup` callback reports the scope size before the first row |
| 4 | A job that has started but reported no progress yet reads as running, not as "Waiting to start" | `PostProcessingJob.run` marks the stage STARTED before the task runs, matching the other job types |
| 5 | Comment norms written down, so the next PR does not repeat the same review round | `.agents/AGENTS.md` gains a rule on comment length and linking to the PR |

### Why de-duplication was both unnecessary and expensive

A classification reaches an occurrence through a plain foreign key, and it reaches a collection through a membership table that holds one row per (capture, collection) pair. Filtering on a single occurrence or a single collection therefore cannot multiply rows, so there was nothing to de-duplicate. The two new scope tests pin exactly that, including the case a reader would worry about: a capture that belongs to several collections still contributes one row per classification.

The cost comes from what `SELECT DISTINCT` has to compare. The selected columns include the `logits` and `scores` arrays, and a production classifier can carry tens of thousands of categories, which puts those two arrays at roughly half a megabyte per row. The database was sorting all of them to prove uniqueness of a column set that already contains the primary key.

De-duplication is also a blocking step, so it delayed the iteration pass as well as the count — no row could be returned until the whole sort finished. Both delays land before the first batch commits, and the first batch commit is where the run reports its first progress.

### Measured, not estimated

Against a live database, on the scope of a run that had been cancelled:

| Query | Time | Result |
|---|---|---|
| Count the scope, with `.distinct()` | 208.0 s | 55,530 |
| Count the scope, without `.distinct()` | 0.5 s | 55,530 |
| Read the first 200 rows, without `.distinct()` | 0.4 s | — |

For context on the remaining setup work, expanding that classifier's category map (29,176 categories) took 10.4 seconds and reading the taxa list took 0.4 seconds. With the de-duplication gone, the work ahead of the first progress report is on the order of ten seconds rather than minutes.

### The stage-status fix, and two corrections it went through

A stage that has started but not yet reported progress reads as "Waiting to start", because `get_status_label` renders a `CREATED` stage that way regardless of progress. Every job type is supposed to move its stage to `STARTED` at the top of `run()`; `PostProcessingJob` was the only one that did not, so its stage sat in `CREATED` until the task's first progress report, which on a large masking run is minutes in.

This took two rounds of review to land correctly, and the audit trail is worth keeping:

- I first cited the cancelled job's stage still being `CREATED` as evidence of the stall. @copilot pointed out a healthy run looked identical, so that was not diagnostic; the evidence that stands is the counter, still at zero eleven minutes in.
- I then fixed it inside `BasePostProcessingTask.update_progress`, guarding the promotion to fire only from `CREATED`. @coderabbitai correctly noted that `Job.update_progress` already promotes on any progress above zero, so my first test at 0.5 was vacuous, and that a per-heartbeat write is the wrong shape for a status transition.
- The takeaway review then placed it correctly: this belongs in `PostProcessingJob.run`, next to `update_status(STARTED)`, exactly as `RegroupEventsJob` and `SourceImageCollectionPopulateJob` already do it. That is one line, it fires before the scope count rather than at the first heartbeat, and it removes the base-class guard entirely. `test_post_processing_stage_is_started_before_the_task_runs` pins it.

### A note on the one `.distinct()` that stays

`ClassMaskingActionForm._algorithms_for_scope` still de-duplicates, and should: it joins from `Algorithm` across a reverse relation to many classifications, so one algorithm genuinely does appear once per classification. It also selects only algorithm columns, so the comparison is cheap.

## Testing

- `ami.ml.post_processing`, `ami.ml.tests` and `ami.jobs` pass, including five new tests.
- `black`, `isort` and `flake8` clean on the touched files.
- The equivalence claim is checked two ways: the new tests assert the scope returns each classification exactly once, and the measurement above confirms both forms of the query return 55,530 on real data.

### CODE REVIEW

File: ami/ml/post_processing/class_masking.py [hunk 1]
```diff
@@ -315,6 +339,10 @@ def run(self) -> None:
         classifications, scope_desc = self._scoped_classifications(config, source_algorithm)
         self.logger.info(f"Applying class masking on {scope_desc} using taxa list {taxa_list.pk}")
 
+        def _on_setup(total: int) -> None:
+            self.update_progress(0.0)
+            self.report_stage_metrics({"classifications_total": total})
+
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/post_processing/class_masking.py:336]: **Claude says:** Correct, and the cause is broader than the callback — thank you, this was a good catch.

Nothing in the post-processing framework ever set a stage status. `PostProcessingJob.run` creates the stage and, after the task returns, marks it SUCCESS; in between, `update_progress` only ever wrote `progress`. So every post-processing run, the small size filter included, was labelled "Waiting to start" for its entire duration.

Fixed in 41e5e05f by having `BasePostProcessingTask.update_progress` mark the stage STARTED, pinned by `test_progress_marks_the_stage_started`. I checked the blast radius before putting it in the shared base class: `JobProgress.is_complete()` gates on the status being final (SUCCESS/FAILURE/REVOKED), and STARTED is no more final than CREATED, so that path is unchanged.

I have also corrected the PR description, which cited the cancelled job's stage still being CREATED as evidence of the stall. As you point out, a healthy run looked identical, so that was not diagnostic. The evidence that stands is the counter, still at zero eleven minutes in.

File: ami/ml/post_processing/tests/test_class_masking.py [hunk 2]
```diff
@@ -486,3 +486,133 @@ def test_reweight_false_winner_identical_scores_differ(self):
...
+
+        for kwargs in (
+            {"source_image_collection_id": self.collection.pk},
+            {"occurrence_id": occurrence.pk},
+        ):
+            with self.subTest(**kwargs):
+                task = ClassMaskingTask(taxa_list_id=taxa_list.pk, algorithm_id=self.algorithm.pk, **kwargs)
+                scoped, _ = task._scoped_classifications(task.config, self.algorithm)
+                self.assertNotIn("DISTINCT", str(scoped.query).upper())
+
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/post_processing/tests/test_class_masking.py:547]: **Claude says:** Agreed — switched to `assertFalse(scoped.query.distinct)` in 41e5e05f. Reading the attribute states the intent directly and does not depend on how the SQL is rendered.

File: ami/ml/post_processing/class_masking.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/post_processing/class_masking.py:336]: **Claude says:** Not applying this one — an empty scope already finishes at 100%.

`PostProcessingJob.run` finalises the stage once the task returns:

```python
task.run()
job.progress.update_stage(cls.key, status=JobState.SUCCESS, progress=1)
```

(`ami/jobs/models.py:912-913`). So for `total == 0` the run ends SUCCESS at 100%, not 0%. The 0.0 written by `_on_setup` is only visible for the moment between the callback and the wrapper finishing, which for an empty scope is immediate.

Reporting 1.0 from the setup callback would also be inaccurate in the non-empty-but-tiny case only by luck of ordering, and it would state that work is complete before any has been done. The stage does now correctly show as running rather than "Waiting to start" — that was a real gap, raised separately by @copilot on this same line and fixed in 41e5e05f.

File: ami/ml/tests.py [hunk 3]
```diff
@@ -1651,6 +1651,37 @@ def test_progress_save_bumps_updated_at_for_reaper(self):
...
+        job.save()
+        self.assertEqual(job.progress.stages[0].status, JobState.CREATED)
+
+        task = SmallSizeFilterTask(job=job, source_image_collection_id=self.collection.pk, size_threshold=0.01)
+        task.update_progress(0.5)
+
+        job.refresh_from_db()
+        stage = job.progress.stages[0]
+        self.assertEqual(stage.status, JobState.STARTED)
+        self.assertEqual(stage.status_label, "50% complete")
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/tests.py]: **Claude says:** You are right, and this one corrected me rather than just the test — thank you.

I had claimed in the previous commit message that nothing ever set a stage status. `Job.update_progress` does, as you say, promoting CREATED to STARTED as soon as progress exceeds zero, and `Job.save()` calls it. So `update_progress(0.5)` was already promoted by the existing code and my test passed with or without the change. Vacuous, exactly as you describe.

The real gap is narrower than I described: progress of exactly zero does not clear the `> 0` bar, and exactly zero is what the setup report introduced by this PR means. Fixed in 3c1d5428 — the test now uses `update_progress(0.0)`, which is the value the setup callback actually produces, and asserts the label reads "0% complete" rather than "Waiting to start". The PR description has been corrected too.

File: ami/ml/post_processing/base.py [hunk 4]
```diff
@@ -74,7 +74,15 @@ def update_progress(self, progress: float):
...
-            self.job.progress.update_stage(self.job.job_type_key, progress=progress)
+            # Imported here because ami.jobs.models imports the post-processing
+            # registry, so a module-level import would be circular.
+            from ami.jobs.models import JobState
+
+            # The stage is marked STARTED here because nothing else does it: the job
+            # wrapper creates the stage and later marks it SUCCESS, so a stage
+            # reporting progress would otherwise still be labelled "Waiting to
+            # start" for the whole run. See #1376.
+            self.job.progress.update_stage(self.job.job_type_key, status=JobState.STARTED, progress=progress)
```
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/post_processing/base.py]: **Claude says:** Agreed, fixed in 3c1d5428. The promotion is now conditional on the stage being in CREATED:

```python
stage = self.job.progress.get_stage(self.job.job_type_key)
status = {"status": JobState.STARTED} if stage.status == JobState.CREATED else {}
self.job.progress.update_stage(self.job.job_type_key, progress=progress, **status)
```

This mirrors the condition `Job.update_progress` already applies to the same transition, so the two cannot disagree. `test_heartbeat_does_not_resurrect_a_finished_stage` covers SUCCESS, FAILURE and REVOKED.

Worth noting the path is reachable without the reaper: `Job.update_progress` promotes a stage to SUCCESS once its progress reaches 1, so any further heartbeat on a completed stage would have demoted it. That is the case the guard closes.

File: ami/ml/tests.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/tests.py]: **Claude says:** Follow-up: the fix moved layers in `d44e00b2` after a structural review. Rather than promoting the stage from inside `update_progress`, `PostProcessingJob.run` now marks the stage STARTED before the task runs, the same way `RegroupEventsJob` and `SourceImageCollectionPopulateJob` already do. That fires before the scope is even counted, and it lets `update_progress` return to its original form with no per-heartbeat status write. The test now drives `PostProcessingJob.run` end to end (`test_post_processing_stage_is_started_before_the_task_runs`) instead of poking `update_progress` directly.

File: ami/ml/post_processing/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mihow) [reply ami/ml/post_processing/base.py]: **Claude says:** Follow-up: this concern is now moot in `d44e00b2`. The per-heartbeat status write is gone entirely — the stage is marked STARTED once, in `PostProcessingJob.run`, before the task starts, matching the other job types. There is no longer a status write on the heartbeat path to guard, so the resurrection risk you flagged cannot arise.
